### PR TITLE
object: do not create service for external object stores

### LINKED ISSUE #11208: CephObjectStore in external mode with SSL-enabled RGW does not work
**Is this a bug report or feature request?**
* Bug Report

I'm trying to configure External mode cluster with SSL enabled for RGW.

In the events I see following error, which seems to be the root cause:
```
failed to reconcile CephObjectStoreUser "<NAMESPACE>/<USERNAME>". failed to create/update object store user "<USERNAME>": failed to get details from ceph object user "<USERNAME>": Get "https://rook-ceph-rgw-<STORENAME>.<NAMESPACE>.svc:443/admin/user?format=json&amp;uid=<USERNAME>": x509: certificate is valid for *.<RGWNODE>, <RGWNODE>, not rook-ceph-rgw-<STORENAME>.<NAMESPACE>.svc
```

**Deviation from expected behavior:**

The certificate SSL/TLS certificate is only valid for specific nodes that contain Ceph RGWs. Attempting to address the RGWs via the Kubernetes Service DNS name won't work.

**Expected behavior:**

Users should not have to add the Kubernetes Service endpoint to their certificates to make Rook able to create users (or buckets). Rook should use the endpooint(s) defined in the CephObjectStore CRD directly.

**How to reproduce it (minimal and precise):**

Create an external CephObjectStore and a CephObjectStoreUser for the store.

**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary

**Logs to submit**:

* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

  To get logs, use `kubectl -n <namespace> logs <pod name>`
  When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
  Read [GitHub documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Cluster Status to submit**:

* Output of krew commands, if necessary

  To get the health of the cluster, use `kubectl rook-ceph health`
  To get the status of the cluster, use `kubectl rook-ceph ceph status`
  For more details, see the [Rook Krew Plugin](https://rook.io/docs/rook/latest-release/Troubleshooting/krew-plugin)

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Storage backend version (e.g. for ceph do `ceph -v`):
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/latest-release/Troubleshooting/ceph-toolbox/#interactive-toolbox)):

### PR DESCRIPTION
**Pull request by BlaineEXE**
External CephObjectStores already have endpoints defined by
spec.gateway.externalRgwEndpoints, and if the external store is
configured with TLS (HTTPS), the store's certificates will likely not
accept connections intended for the Service endpoint Rook creates. Some
users might not be able to easily add the service endpoint to their
certificates. Therefore, don't even bother creating a Service for
external clusters.

This does introduce a few issues. The Service seems to have been
initially created to allow multiple external RGW endpoints to be
addressable via a single address in Rook. For all connections to an
external CephObjectStore with multiple endpoints, simply choose an
endpoint at random. Random selection will prevent Rook from failing to
create buckets or users on an external store if one of the external
store's endpoints fails.

The latest OBC library (lib-bucket-provisioner) allows updating the
endpoints on ObjectBuckets after they are created. This allows Rook
users to change endpoints on external CephObjectStores without breaking
all existing OBCs. It requires implementation of the new GetUserID()
library call, requires updating Provision() and Grant() calls to be
idempotent, and it requires removing the Update() call.

Signed-off-by: Blaine Gardner <blaine.gardner@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #11208

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: deploy/examples/crds.yaml [hunk 1]
```diff
@@ -11558,42 +11558,14 @@ spec:
...
-                              namespace:
-                                description: 'Namespace of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/'
-                                type: string
-                              resourceVersion:
-                                description: 'Specific resourceVersion to which this reference is made, if any. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#concurrency-control-and-consistency'
-                                type: string
-                              uid:
-                                description: 'UID of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#uids'
-                                type: string
-                            type: object
```
Comment (BlaineEXE) [inline deploy/examples/crds.yaml:11596]: In order to make the endpoints simpler for us to manage, I created our own copy of the type that omits the `NodeName` and `TargetRef` (`ObjectRef`) fields since I really doubt those are in use from users. But if any users do use these, these changes are breaking.

File: pkg/apis/ceph.rook.io/v1/object.go [hunk 2]
```diff
@@ -113,3 +113,12 @@ func (c *CephObjectStore) GetStatusConditions() *[]Condition {
...
 }
+
+// String returns an addressable string representation of the EndpointAddress.
+func (e *EndpointAddress) String() string {
+	// hostname is easier to read, and it is probably less likely to change, so prefer it over IP
+	if e.Hostname != "" {
+		return e.Hostname
+	}
+	return e.IP
+}
```
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/object.go:124]: Using a pared down version of the `EndpointAddress` allows us to more easily stringify the endpoint. It's either the specified hostname or IP. Nothing fancy.

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 3]
```diff
@@ -1440,6 +1440,22 @@ type GatewaySpec struct {
...
+// EndpointAddress is a tuple that describes a single IP address or host name. This is a subset of
+// Kubernetes's v1.EndpointAddress.
+// +structType=atomic
+type EndpointAddress struct {
+	// The IP of this endpoint.
+	// May not be loopback (127.0.0.0/8), link-local (169.254.0.0/16),
+	// or link-local multicast ((224.0.0.0/24).
+	// IPv6 is also accepted but not fully supported on all platforms. Also, certain
+	// kubernetes components, like kube-proxy, are not IPv6 ready.
+	// TODO: This should allow hostname or IP, See #4447.
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:1452]: Is this TODO still needed? There is a separate IP and hostname field already. #4447 doesn't look related either.

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 4]
```diff
@@ -1440,6 +1440,22 @@ type GatewaySpec struct {
 	HostNetwork *bool `json:"hostNetwork,omitempty"`
 }
 
+// EndpointAddress is a tuple that describes a single IP address or host name. This is a subset of
+// Kubernetes's v1.EndpointAddress.
+// +structType=atomic
+type EndpointAddress struct {
+	// The IP of this endpoint.
+	// May not be loopback (127.0.0.0/8), link-local (169.254.0.0/16),
```
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:1448]: Are all these comments relevant? Seems like we shouldn't really care what type of address is given as long as it is reachable.

File: pkg/operator/ceph/object/controller.go [hunk 5]
```diff
@@ -368,22 +368,20 @@ func (r *ReconcileCephObjectStore) reconcileCreateObjectStore(cephObjectStore *c
...
 
-		// RECONCILE ENDPOINTS
-		// Always add the endpoint AFTER the service otherwise it will get overridden
-		logger.Info("reconciling external object store endpoint")
-		err = cfg.reconcileExternalEndpoint(cephObjectStore)
+		// deleting the service also deletes the endpoint(s), so deleting them isn't necessary
+		err := k8sutil.DeleteService(cfg.clusterInfo.Context, cfg.context.Clientset, service.Namespace, service.Name)
 		if err != nil {
-			return r.setFailedStatus(k8sutil.ObservedGenerationNotAvailable, namespacedName, "failed to reconcile external endpoint", err)
+			logger.Warningf("continuing after non critical error. "+
```
Comment (travisn) [inline pkg/operator/ceph/object/controller.go:379]: If the service doesn't exist we don't need to log this, right?

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/object/controller.go (hunk 5 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/object/controller.go:379]: If it doesn't exist, there will be no error from the called function. This is in case of some other error.

**Comment by parth-gr** [comment]
Is this for external clusters, or  ExternalRgwEndpoints ? 
For enabling TLS

File: pkg/apis/ceph.rook.io/v1/types.go [hunk 6]
```diff
@@ -1440,6 +1440,18 @@ type GatewaySpec struct {
 	HostNetwork *bool `json:"hostNetwork,omitempty"`
 }
 
+// EndpointAddress is a tuple that describes a single IP address or host name. This is a subset of
+// Kubernetes's v1.EndpointAddress.
+// +structType=atomic
+type EndpointAddress struct {
+	// The IP of this endpoint.
+	// +optional
+	IP string `json:"ip" protobuf:"bytes,1,opt,name=ip"`
```
Comment (leseb) [inline pkg/apis/ceph.rook.io/v1/types.go:1449]: Are the protobuf annotations really necessary here?

**Comment by leseb** [comment]
> ExternalRgwEndpoints

Well AFAIR `ExternalRgwEndpoints` is external cluster since we add rgw endpoints that are not managed by Rook.

**Review by leseb — COMMENTED** [review]
I left some more questions here https://github.com/rook/rook/issues/11208.

**Comment by BlaineEXE** [comment]
Adding changes from OBC library present here: https://github.com/kube-object-storage/lib-bucket-provisioner/pull/224

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 6 above)
Comment (BlaineEXE) [reply pkg/apis/ceph.rook.io/v1/types.go:1449]: I'm not sure, but it seemed reasonable to keep them to make sure the CRD that gets generated is compatible with the old type.

File: go.mod [hunk 7]
```diff
@@ -292,3 +292,5 @@ exclude (
 	k8s.io/client-go v9.0.0+incompatible
 	k8s.io/client-go v9.0.0-invalid+incompatible
 )
+
+replace github.com/kube-object-storage/lib-bucket-provisioner => github.com/BlaineEXE/lib-bucket-provisioner v0.0.0-20221102211652-aa9aebb5b74f
```
Comment (travisn) [inline go.mod:296]: planned for revert I assume

File: pkg/apis/ceph.rook.io/v1/types.go (hunk 6 above)
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:1451]: either way should be fine, they should be ignored since we're not using protobuf internally with rook

File: pkg/operator/ceph/object/bucket/provisioner.go [hunk 8]
```diff
@@ -87,32 +96,25 @@ func (p Provisioner) Provision(options *apibkt.BucketOptions) (*bktv1alpha1.Obje
...
-		p.deleteOBCResourceLogError("")
 		return nil, err
 	}
 
 	// create the bucket
 	err = s3svc.CreateBucket(p.bucketName)
 	if err != nil {
-		err = errors.Wrapf(err, "error creating bucket %q", p.bucketName)
-		logger.Errorf(err.Error())
-		p.deleteOBCResourceLogError("")
```
Comment (travisn) [inline pkg/operator/ceph/object/bucket/provisioner.go:99]: I remember thinking this was less than ideal to delete the OBCs during failure in every failure path, but where is it being cleaned up now? Or is it not needed with the new idempotent approach?

File: pkg/operator/ceph/object/bucket/provisioner.go [hunk 9]
```diff
@@ -136,29 +138,27 @@ func (p Provisioner) Grant(options *apibkt.BucketOptions) (*bktv1alpha1.ObjectBu
...
+	p.accessKeyID, p.secretAccessKey, err = p.createCephUser(options.UserID)
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrap(err, "Provision: can't create ceph user")
 	}
 
-	// need to quota into -1 for restricting creation of new buckets in rgw
-	restrictBucketCreation := -1
+	// restrict creation of new buckets in rgw
+	restrictBucketCreation := 0
```
Comment (travisn) [inline pkg/operator/ceph/object/bucket/provisioner.go:148]: Changing the quota from -1 to 0 is effectively still the same, right? Or was there some side effect of -1?

File: pkg/operator/ceph/object/bucket/rgw-handlers.go [hunk 10]
```diff
@@ -65,24 +47,10 @@ func (p *Provisioner) createCephUser(username string) (accKey string, secKey str
...
-	// radosgw-admin will just return the existing user.
-	// when notUnique == true, the loop breaks and `name` contains the latest generated name
-	for i := 0; notUnique && i < maxTries; i++ {
-		genName = fmt.Sprintf("%s-%s", prefix, randomString(genUserLen))
-		if notUnique, err = p.userExists(genName); err != nil {
-			return "", err
-		}
-	}
-	return genName, nil
+func (p *Provisioner) genUserName(obcName, obcNamespace string) string {
```
Comment (travisn) [inline pkg/operator/ceph/object/bucket/rgw-handlers.go:50]: much simpler now!

**Review by travisn — COMMENTED** [review]
LGTM, just a few questions

File: pkg/operator/ceph/object/bucket/provisioner.go (hunk 9 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/object/bucket/provisioner.go:148]: It's actually a behavior fix I noticed. -1 actually disables the quota. 0 is the value that specifies "no new buckets". The commit that added this was pretty old, so it's possible there was some confusion at the time.

File: pkg/operator/ceph/object/bucket/provisioner.go (hunk 8 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/object/bucket/provisioner.go:99]: It's not needed with the idempotent approach. L-B-P has generated the bucket name before Provision() for a while, and now with GetUserID being deterministic, there shouldn't be any risk of leaving dangling buckets/users with random names. We can expect to clean up when L-B-P calls Delete().

File: go.mod (hunk 7 above)
Comment (BlaineEXE) [reply go.mod:296]: Yes. We need to merge the L-B-P changes before this PR.

Comment (BlaineEXE) [reply go.mod]: That's done now. If things are looking good, I'll merge once it seems like CI is stable.

Comment (BlaineEXE) [reply go.mod]: Note to self: I also want to go through the upgrade test after my latest changes. It went well a few days ago, but I should double check.

**Comment by BlaineEXE** [comment]
@travisn upgrade testing went fine after I made a small change. Github is still showing your request for changes.

**Comment by parth-gr** [comment]
@BlaineEXE in your testing did you have a chance to test end-to-end external cluster creation with ODF and SSL TLS enabled?

**Review by travisn — APPROVED** [review]
